### ISSUE TITLE
TypeScript typings

### ISSUE THREAD

**Issue by feimosi** [plopjs/plop#143]
Plop is a great tool already, but do you consider adding TypeScript typings? I've noticed `node-plop` has them, but they're not being exported so I can't reuse them in my code. Also, they differ a little bit from this CLI.

In case anyone needs a temporary solution, here's a simple `.d.ts`:

```ts
import inquirer = require('inquirer');

export interface Plop {
  getGenerator(name: string): PlopGenerator;
  setGenerator(name: string, config: PlopGenerator): PlopGenerator;
}

interface PlopGenerator{
  description: string;
  prompts: inquirer.Question[];
  actions: ActionConfig[];
}

interface ActionConfig{
  type: string;
  path: string;
  templateFile: string;
  force?: boolean;
  data?: object;
  abortOnFail?: boolean;
}

interface PlopCfg{
  force: boolean;
  destBasePath: string;
}
```

**Comment by amwmedia** [comment]
If you'd like to submit a PR, I'd be open to adding them. I don't currently have a lot of TypeScript experience, so I'm not sure of the best way to provide these in a usable way. Sounds like it might be as simple as a `.d.ts` file and a line or 2 in the documentation. 👍

**Comment by crutchcorn** [comment]
I'll set a reminder for myself to take a look at this this weekend

**Comment by feimosi** [comment]
@crutchcorn thanks for taking this over. I've been pretty busy lately

**Comment by crutchcorn** [comment]
Sorry folks, meant to get to this sooner. I won't have a ton of time to look at it for a few days (spending time with family), but I've done a bit of look into this.

Keep in mind that the definitions are currently in `node-plop`, so I'm not sure how we want to go about tracking these issues.

Progress I've made thus far is found here:
https://github.com/crutchcorn/node-plop/commit/9b13d3df60314ee5f00fa3b9e5dfb674079fe1f7

I think there's a few things going on:
1) The definitions are not nearly as strict as I think makes sense. This is why @feimosi mentioned there not being aligned types. I have gone ahead and added new types where things were missing and also went ahead and improved type strictness anyway. 
2) The types that might be useful for `plop` users are not being exported (in `node-plop`, where they currently reside), I've gone ahead and added exports to things I think might be useful for the plopfile
3) They are not being exported in `plop` at all. This commit does not handle that, but I need the changes to be finalized (by myself) and merged (by upstream when I've finished) before making a PR to `plop` that adds the `index.d.ts` file to export the types from `node-vibrant`. I will also add docs on the types and TS stuff in general when doing this step

**Comment by crutchcorn** [comment]
I've opened ammwmedia/node-plop#97 as a continuation to what I've found. Once this PR is handled, I'll move forward with the work to expose these typings in `plop` itself and add some docs on their usefulness/etc

**Comment by amwmedia** [comment]
Thanks for all your work on this. My apologies for the delay in my end. I seem to get sporadic chunks of time to work on plop.

**Comment by crutchcorn** [comment]
No worries at all, I totally understand time restraints. Hopefully I haven't been spamming your notifications too hard, just trying to keep communication open and clear.

**Comment by crutchcorn** [comment]
Actually, @amwmedia - could we get an updated version of `node-plop` up on `npm` with the updated typings so that I can rely on that version of `node-plop`  to export some typings in the `plop` package?

**Comment by amwmedia** [comment]
this was updated and pushed out to npm 👍

**Comment by crutchcorn** [comment]
I still need to export the typings from the main `plop` package and write some docs. I'll go ahead and do that now :)

**Comment by amwmedia** [comment]
Ah, I'll re-open till that work is complete. Thanks for your help on this. I am, admittedly, not a TypeScript expert.

**Comment by amwmedia** [comment]
released with v2.5.0
